Sources/OpenCombine/Publishers/Publishers.Decode.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  Publishers.Decode.swift
+//  
+//
+//  Created by Joseph Spadafora on 6/21/19.
+//
+
+
+extension Publishers {
+    
+    public struct Decode<Upstream, Output, Coder> : Publisher where Upstream : Publisher, Output : Decodable, Coder : TopLevelDecoder, Upstream.Output == Coder.Input {
+        
+        /// The kind of errors this publisher might publish.
+        ///
+        /// Use `Never` if this `Publisher` does not publish errors.
+        public typealias Failure = Error
+        
+        public let upstream: Upstream
+        
+        let decoder: Coder
+        
+        public init(upstream: Upstream, decoder: Coder) {
+            self.upstream = upstream
+            self.decoder = decoder
+        }
+        
+        /// This function is called to attach the specified `Subscriber` to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<S: Subscriber>(subscriber: S)
+            where Failure == S.Failure, Output == S.Input
+        {
+            let decodeSubscriber = _Decode<Upstream, S, Coder>(downstream: subscriber, decoder: decoder)
+            upstream.receive(subscriber: decodeSubscriber)
+        }
+    }
+}
+
+private final class _Decode<Upstream: Publisher, Downstream: Subscriber, Coder: TopLevelDecoder>:
+    Subscriber,
+    CustomStringConvertible,
+    CustomReflectable,
+Subscription where Downstream.Input: Decodable, Coder.Input == Upstream.Output {
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+    typealias Output = Downstream.Input
+    
+    private let _decoder: Coder
+    private var _downstream: Downstream
+    private var _upstreamSubscription: Subscription?
+    private var _demand: Subscribers.Demand = .none
+    
+    var description: String { return "Decode" }
+    
+    var customMirror: Mirror { return Mirror(self, children: EmptyCollection()) }
+    
+    init(downstream: Downstream, decoder: Coder) {
+        self._downstream = downstream
+        self._decoder = decoder
+    }
+    
+    func receive(subscription: Subscription) {
+        _upstreamSubscription = subscription
+        subscription.request(.max(1))
```

Comment:
Are you sure we need to request `.max(1)` here? In `DropWhile` I've checked that this is what Apple's Combine does, and it made sense because we wanted to skip some elements no matter how much the downstream subscriber has requested.

Here is the test that ensured this behavior for `DropWhile`:
https://github.com/broadwaylamb/OpenCombine/blob/17f9b0a29975106d2f51d5d9b7858c5b1e612019/Tests/OpenCombineTests/PublisherTests/DropWhileTests.swift#L90-L129

I haven't checked this for `Decode`/`Encode` yet, so I'd be great if you could write a similar test for the `Decode`/`Encode` operator and run it in compatibility mode (and tweak your implementation accordingly). If you feel like it'd require too deep dive that you don't have much time for, it's perfectly fine, I'll do it myself.

Reply:
@broadwaylamb It seems like you asked MortyMerr for a very similar demand test.  Is there any way that we can abstract the demand so that we aren't testing the same thing and writing the same code inside every publisher?

Reply:
Keep in mind that we have to be able to run the tests in exactly the same way against Apple's Combine — that's like 80% why we're writing those tests at all.

We could make a test function generic over some `Publisher`, but we can't use it to test all the publishers. For example, the operators that have map-like semantics (`Map`, `Encode`, `Decode`, `Scan`), i. e. those that preserve the number and periodicity of published events, do not change the demand. For those, we can write a generic test.

However, operators like `DropWhile`, `Filter` etc. may request the different number of elements than their downstream, so we can't naively write a fully generic test that could test all the operators without modification, it is likely better to write ad-hoc tests.

But in my opinion, we should write an ad-hoc test first, and only then make it generic.

Reply:
@broadwaylamb Okay, I'm fine with writing an ad-hoc test.  It looks like the tracking subscriber only works for something that maps to Int values.  

Can you clarify this a bit for me: 
>I haven't checked this for Decode/Encode/Map yet, so I'd be great if you could write a similar test for the Decode/Encode/Map operator and run it in compatibility mode (and tweak your implementation accordingly). If you feel like it'd require too deep dive that you don't have much time for, it's perfectly fine, I'll do it myself.

For example, how did you check that `DropWhile` had a request of 1?  As far as compatibility mode, what is the exact process and what am I looking for to know if I've been successful or not?